Add iframe search with debounce

rzhao271 opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #96307 by adding an iframe search prototype. It uses https://github.com/electron/electron/pull/28274 to search in the iframe, which results in a loss of focus during the search. Therefore, I added an event to re-focus the inputbox after find results are returned, and also added a debounce so that the focus-changing behaviour of the find call is not too obvious.

Preview:
![recording (31)](https://user-images.githubusercontent.com/7199958/121382090-1df6eb80-c8fb-11eb-920a-f3a4cb22da5a.gif)

Issues:
- The jumping search problem remains (https://github.com/microsoft/vscode/issues/96307#issuecomment-848137542). This issue can be fixed by using a BrowserView, in which case, the debounce can be taken off as well (https://github.com/microsoft/vscode/issues/96307#issuecomment-844362867).

rzhao271 commented:
An issue with the CI seems to be that it's not taking in the proper Electron.
Even locally, I get the same compile errors, but with the right Electron, the code still runs properly.

rzhao271 commented:
Verified that in the case of an IME (e.g. for Mandarin), the search must be made explicit, e.g. by pressing enter, otherwise focus is lost from the IME. CC @rebornix 

Meanwhile when using NVDA, the readout is very different than in stable, or in the Monaco editor. In the editor, NVDA starts reading the document from the active match. In stable, NVDA only reads what is being typed in the find box. With this PR, when searching in an iframe, NVDA _sometimes_ says "virtual document" and starts reading the iframe from the begnning, due to focus being lost from the inputbox for long enough to trigger the screenreader. Other times, NVDA only reads what is typed. @isidorn I'm wondering how to make the search behaviour more accessible here?

mjbvz commented on src/vs/platform/webview/electron-main/webviewMainService.ts:
```diff
@@ -57,4 +62,43 @@ export class WebviewMainService extends Disposable implements IWebviewManagerSer
+		return frame;
+	}
+
+	override dispose() {
```
Make sure to also call `super.dispose()`

mjbvz commented on src/vs/workbench/contrib/webview/electron-sandbox/iframeWebviewElement.ts:
```diff
@@ -29,6 +36,12 @@ export class ElectronIframeWebview extends IFrameWebview {
 
 	private readonly _webviewKeyboardHandler: WindowIgnoreMenuShortcutsManager;
```
Add `implements WebviewFindDelegate` to make the contract more explicit

mjbvz commented on src/vs/workbench/contrib/webview/electron-sandbox/iframeWebviewElement.ts:
```diff
@@ -29,6 +36,12 @@ export class ElectronIframeWebview extends IFrameWebview {
+	private _findStarted: boolean = false;
+
+	private readonly _webviewMainService: IWebviewManagerService;
+	private readonly _iframeDelayer = new Delayer<void>(200);
```
[nit] make sure to register the disposable

bpasero replied:
```suggestion
	private readonly _iframeDelayer = this._register(new Delayer<void>(200));
```

mjbvz commented on src/vs/platform/webview/electron-main/webviewMainService.ts:
```diff
@@ -3,17 +3,22 @@
 
 	declare readonly _serviceBrand: undefined;
 
+	private frameHandlers = new Map<string, WebFrameMain>();
```
Does this need to be a map? I don't see cleanup logic for deleting entries when a webview is disposed of so I'm worried the `WebFrameMain` could be leaked . You may be able to avoid if this was a `set` (or, even better, is that a way you can avoid having state all together?)

rzhao271 replied:
Changed around the logic to not require a map here.

mjbvz reviewed (changes requested):
Code changes look good overall. Nice work! Just a few minor comments

bpasero commented on product.json:
```diff
@@ -23,6 +23,7 @@
 	"reportIssueUrl": "https://github.com/microsoft/vscode/issues/new",
 	"urlProtocol": "code-oss",
 	"webviewContentExternalBaseUrlTemplate": "https://{{uuid}}.vscode-webview.net/{{quality}}/{{commit}}/out/vs/workbench/contrib/webview/browser/pre/",
+	"electronRepository": "microsoft/vscode-electron-prebuilt",
```
Make sure to remove this before merging.

bpasero commented on src/vs/platform/webview/electron-main/webviewMainService.ts:
```diff
@@ -57,4 +62,43 @@ export class WebviewMainService extends Disposable implements IWebviewManagerSer
+
+	public async findInFrame(windowId: WebviewWindowId, frameName: string, text: string, options: { findNext?: boolean, forward?: boolean }): Promise<void> {
+		const frame = this.getFrameByName(windowId, frameName);
+		frame.findInFrame(text, {
```
I think these methods do not exist in the public `electron.d.ts` so you will have to do function probing and casting similar to what I have done here:

https://github.com/microsoft/vscode/blob/2cb7b42d98e44fade2067f06f61c3cf5d1defd58/src/vs/code/electron-main/app.ts#L150-L165

bpasero commented on src/vs/platform/webview/electron-main/webviewMainService.ts:
```diff
@@ -57,4 +62,43 @@ export class WebviewMainService extends Disposable implements IWebviewManagerSer
+
+	public async stopFindInFrame(windowId: WebviewWindowId, frameName: string, options: { keepSelection?: boolean }): Promise<void> {
+		const frame = this.getFrameByName(windowId, frameName);
+		frame.stopFindInFrame(options.keepSelection ? 'keepSelection' : 'clearSelection');
```
Same here.

bpasero commented on src/vs/platform/webview/electron-main/webviewMainService.ts:
```diff
@@ -57,4 +62,43 @@ export class WebviewMainService extends Disposable implements IWebviewManagerSer
+					this._onFoundInFrame.fire(result);
+				}
+			});
+			this.frameHandlers.set(frameName, event);
```
I am missing some code that clears the frame handler when done, does this end up piling events up, then it sounds like a potential memory leak?

rzhao271 replied:
Changed around the code to not require `frameHandlers` variable.

bpasero reviewed (changes requested):
👏 👏 👏 

Gave some feedback inline. Also found one potential issue: I cannot hit `Cmd+X` to cut/clear the input of the find input box. Also `Cmd+C` and `Cmd+V` do not seem to work there, maybe something that is still missing?

isidorn commented:
@rzhao271 thanks for the ping. I see there are already quite some discussions here so for accessibility I think the best would be that once we merge this in you create an issue for accessibility and we look into that together. Once this is merged it will also be easier for me to try it out.
Let me know if that works for you? Thanks

rzhao271 commented:
@bpasero I can confirm that the cut, copy, and paste commands don't work, but I'm not sure why, yet. I notice there's this event handler that is run when I type Ctrl+X, but I'm not sure if it interferes in any way https://github.com/microsoft/vscode/blob/main/src/vs/base/browser/ui/findinput/findInput.ts#L215. @mjbvz do you have any insights on this?

Also, with the IME issue, if we switch webviews to iframes on Insiders, searching will fail for anyone using an IME. Therefore, I think we can merge this to Insiders, but I wouldn't set this search to be the default right now.

rzhao271 commented:
Another issue: OSS seems to crash when I do a search, but then change the content in the markdown file.

rzhao271 commented:
The match colour alternating between yellow and orange is an artifact of the find-in-frame implementation:
![recording (33)](https://user-images.githubusercontent.com/7199958/121607842-9e097800-ca05-11eb-975d-d570705b318c.gif)

The crash described in https://github.com/microsoft/vscode/pull/125856#issuecomment-859111525 is irregular, though, and requires I make a crashdump.

bpasero reviewed (approved):
🚢  it

bpasero commented:
> Another issue: OSS seems to crash when I do a search in the preview view, but then change the content in the source markdown file.

What are the steps to get to this crash? When you say OSS you mean running with our internal builds?

rzhao271 commented:
I confirmed the iframe search crash occurs around the time the `html` method of the base webview is called, which posts the following message:

```
[8916:0618/120706.592:INFO:CONSOLE(73)] "posting message", source: file:///C:/Users/raymondzhao/work/vscode-11/out/vs/workbench/contrib/webview/electron-sandbox/iframeWebviewElement.js (73)
[8916:0618/120706.592:INFO:CONSOLE(74)] "channel: content", source: file:///C:/Users/raymondzhao/work/vscode-11/out/vs/workbench/contrib/webview/electron-sandbox/iframeWebviewElement.js (74)
[8916:0618/120706.592:INFO:CONSOLE(75)] "data: {"contents":"<!DOCTYPE html>\n\t\t\t<html style=\"--markdown-font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif; --markdown-font-size: 14px; --markdown-line-height: 1.6;\">\n\t\t\t<head>\n\t\t\t\t<meta http-equiv=\"Content-type\" content=\"text/html;charset=UTF-8\">\n\t\t\t\t<meta http-equiv=\"Content-Security-Policy\" content=\"default-src 'none'; img-src 'self' https://*.vscode-webview.net https: data:; media-src 'self' https://*.vscode-webview.net https: data:; script-src 'nonce-1624043226590590'; style-src 'self' https://*.vscode-webview.net 'unsafe-inline' https: data:; font-src 'self' https://*.vscode-webview.net https: data:;\">\n\t\t\t\t<meta id=\"vscode-markdown-preview-data\"\n\t\t\t\t\tdata-settings=\"{&quot;source&quot;:&quot;file:///c%3A/Users/raymondzhao/work/emmet-work/testworkspace3/test.md&quot;,&quot;line&quot;:0,&quot;lineCount&quot;:1,&quot;scrollPreviewWithEditor&quot;:true,&quot;scrollEditorWithPreview&quot;:true,&quot;doubleClickToSwitchToEditor&quot;:true,&quot;disableSecurityWarnings&quot;:false,&quot;webviewResourceRoot&quot;:&quot;https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/emmet-work/testworkspace3/test.md&quot;}\"\n\t\t\t\t\tdata-strings=\"{&quot;cspAlertMessageText&quot;:&quot;Some content has been disabled in this document&quot;,&quot;cspAlertMessageTitle&quot;:&quot;Potentially unsafe or insecure content has been disabled in the Markdown preview. Change the Markdown preview security setting to allow insecure content or enable scripts&quot;,&quot;cspAlertMessageLabel&quot;:&quot;Content Disabled Security Warning&quot;}\"\n\t\t\t\t\tdata-state=\"{&quot;resource&quot;:&quot;file:///c%3A/Users/raymondzhao/work/emmet-work/testworkspace3/test.md&quot;,&quot;line&quot;:0,&quot;imageInfo&quot;:[],&quot;resourceColumn&quot;:1,&quot;locked&quot;:false}\">\n\t\t\t\t<script src=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features/media/pre.js\" nonce=\"1624043226590590\"></script>\n\t\t\t\t<link 
rel=\"stylesheet\" type=\"text/css\" href=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features/media/markdown.css\">\n<link rel=\"stylesheet\" type=\"text/css\" href=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features/media/highlight.css\">\n<link rel=\"stylesheet\" type=\"text/css\" href=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-math/node_modules/katex/dist/katex.min.css\">\n<link rel=\"stylesheet\" type=\"text/css\" href=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-math/preview-styles/index.css\">\n\t\t\t\n\t\t\t<style>\n</style>\n\n\t\t\t\t<base href=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/emmet-work/testworkspace3/test.md\">\n\t\t\t</head>\n\t\t\t<body class=\"vscode-body scrollBeyondLastLine wordWrap 
showEditorSelection\">\n\t\t\t\t<p data-line=\"0\" class=\"code-line\">Some texta</p>\n\n\t\t\t\t<div class=\"code-line\" data-line=\"1\"></div>\n\t\t\t\t<script async\n\t\t\t\tsrc=\"https://file%2B.vscode-resource.vscode-webview.net/c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features/media/index.js\"\n\t\t\t\tnonce=\"1624043226590590\"\n\t\t\t\tcharset=\"UTF-8\"></script>\n\t\t\t</body>\n\t\t\t</html>","options":{"enableFindWidget":true,"allowScripts":true,"localResourceRoots":[{"$mid":1,"fsPath":"c:\\Users\\raymondzhao\\work\\vscode-11\\extensions\\markdown-language-features","_sep":1,"external":"file:///c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features","path":"/c:/Users/raymondzhao/work/vscode-11/extensions/markdown-language-features","scheme":"file"},{"$mid":1,"fsPath":"c:\\Users\\raymondzhao\\work\\vscode-11\\extensions\\markdown-math","_sep":1,"external":"file:///c%3A/Users/raymondzhao/work/vscode-11/extensions/markdown-math","path":"/c:/Users/raymondzhao/work/vscode-11/extensions/markdown-math","scheme":"file"},{"$mid":1,"external":"file:///c%3A/Users/raymondzhao/work/emmet-work/testworkspace3","path":"/c:/Users/raymondzhao/work/emmet-work/testworkspace3","scheme":"file"}]},"state":"{\"resource\":\"file:///c%3A/Users/raymondzhao/work/emmet-work/testworkspace3/test.md\",\"line\":0,\"imageInfo\":[],\"resourceColumn\":1,\"locked\":false,\"scrollProgress\":0}","cspSource":"https://*.vscode-webview.net"}", source: file:///C:/Users/raymondzhao/work/vscode-11/out/vs/workbench/contrib/webview/electron-sandbox/iframeWebviewElement.js (75)
```

I tried adding some code that stopped the find before that message was posted, but the crash continued to occur.

bpasero commented:
👍 , I have extracted https://github.com/microsoft/vscode/issues/126729 for the crash. I found a second case where I can reproduce a crash.

rzhao271 commented:
I have rebased the changes on top of Deepak's FindRequestManager fix, which now avoids the crash described in #126729.

I also added on a commit that adds on `onDidStopFind` event, because the find could be stopped by a content update, in which case, focus should stay in the editor (previously, focus would always go back to the search box).

rzhao271 commented on src/vs/workbench/contrib/codeEditor/browser/find/simpleFindWidget.ts:
```diff
@@ -68,11 +69,16 @@ export abstract class SimpleFindWidget extends Widget {
-			this._delayedUpdateHistory();
-		});
+		this._register(this._findInput.onInput((e) => {
+			if (checkImeCompletionState && !this._findInput.isImeSessionInProgress) {
```
Note to self before merge: I'm not sure whether this condition is correct.
I think it should be that if we're checking for the completion state, then we check if the session is not in progress.

rzhao271 replied:
Fixed in latest commit. Also removed empty lines in this block.
